src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp

```diff
@@ -0,0 +1,116 @@
+///
+/// DataBox changes:
+/// - Adds: nothing
+/// - Removes: nothing
+/// - Modifies: nothing
+template <typename InterpolationTargetTag>
+struct InterpolateToTarget {
+  template <typename DbTags, typename Metavariables, typename... InboxTags,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTags>&&> apply(
+      db::DataBox<DbTags>& box, tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      Parallel::ConstGlobalCache<Metavariables>& cache,
+      const ArrayIndex& array_index, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    static constexpr size_t dim = Metavariables::volume_dim;
+
+    // Get element logical coordinates.
+    const auto& block_logical_coords =
+        get<Vars::PointInfoTag<InterpolationTargetTag, dim>>(
+            db::get<Tags::InterpPointInfo<Metavariables>>(box));
+    const std::vector<ElementId<dim>> element_ids{{array_index}};
+    const auto element_coord_holders =
+        element_logical_coordinates(element_ids, block_logical_coords);
+
+    // There is exactly one element_id in the list of element_ids.
+    if (element_coord_holders.count(element_ids[0]) == 0) {
+      // There are no points in this element, so we don't need
+      // to do anything.
+      return std::forward_as_tuple(std::move(box));
+    }
+
+    // There are points in this element, so interpolate to them and
+    // send the interpolated data to the target.  This is done
+    // in several steps:
+    const auto& element_coord_holder = element_coord_holders.at(element_ids[0]);
+
+    // 1. Create a new DataBox that contains
+    // InterpolationTargetTag::compute_items_on_source
+    auto new_box = db::create_from<
+        db::RemoveTags<>, db::AddSimpleTags<>,
+        db::AddComputeTags<
+            typename InterpolationTargetTag::compute_items_on_source>>(
+        std::move(box));
+
+    // 2. Set up local variables to hold vars_to_interpolate and fill it
+    const auto& mesh = db::get<domain::Tags::Mesh<dim>>(new_box);
+    Variables<typename InterpolationTargetTag::vars_to_interpolate_to_target>
+        local_vars(mesh.number_of_grid_points());
+
+    tmpl::for_each<
+        typename InterpolationTargetTag::vars_to_interpolate_to_target>(
+        [&new_box, &local_vars ](auto x) noexcept {
+          using tag = typename decltype(x)::type;
+          get<tag>(local_vars) = db::get<tag>(new_box);
+        });
+
+    // 3. Set up interpolator
+    intrp::Irregular<dim> interpolator(
+        mesh, element_coord_holder.element_logical_coords);
+
+    // 4. Interpolate and send interpolated data to target
+    auto& receiver_proxy = Parallel::get_parallel_component<
+        InterpolationTarget<Metavariables, InterpolationTargetTag>>(cache);
+    Parallel::simple_action<
+        Actions::InterpolationTargetVarsFromElement<InterpolationTargetTag>>(
+        receiver_proxy,
+        std::vector<Variables<
```

Comment:
What do you mean by copying twice? I see one "obvious" copy, the one in the for_each that copies each tensor into a Variables.  Is the other copy you are referring to the call to the simple_action, which copies its arguments?  Or do you mean that the std::vector in line 100 is being filled by copying the result of interpolator.interpolate instead of moving that result (in which case I should construct the vector differently on line 100)?

Are you suggesting to add a new member function to interp::Irregular that takes a Tensor instead of a Variables, so that the copy in the for_each can be eliminated?

Reply:
Sorry, there was some miscommunication between Larry and I. There is only one additional copy being done. The question is: if the interpolant would support interpolating a `Tensor` by `not_null`, would it be more efficient to just do the interpolation directly rather than copying into a buffer, then interpolating. There are currently 2 memory allocations, which could be reduced to one, but that's unlikely to be an issue.

Reply:
Maybe.  If the interpolant could interpolate a Tensor by not_null, then there would be one
fewer memory allocation.  But then the interpolator would need to be called for each Tensor individually, instead of doing the entire Variables as a single matrix operation.  That is, it would be several matrix-vector multiplies instead of one matrix-matrix multiply.

Reply:
Discussion in SpECTRE call: filing an issue and deferring is fine

Reply:
Did you file an issue for this already @markscheel?

Reply:
Yes, #2271